### ISSUE TITLE
Allow constraint on property and subproperty if constraint on subproperty is more restrictive than on superproperty

### ISSUE THREAD

**Issue by mattssp** [dhlab-basel/Knora#1226]
https://github.com/dhlab-basel/Knora/issues/978 introduced a new consistency check that disallows a class to have a cardinality restriction on a property and on its subproperty. I ran into this while trying to use the nie-ine-ontologies as shared ontologies in my project (see https://github.com/nie-ine/Ontologies/issues/7). These ontologies are making heavy use of this type of restriction.

I believe that there are valid reasons to have restrictions on properties and on its subproperties. If you define an abstract concept that has an abstract property (e.g. a creation that can have 1 or more creators) and you are subclassing this, eg. having a sculpture that has 1 or more sculptors and a text that has 1 or more authors, then you would probably want to use this method. 

In the case discussed in https://github.com/dhlab-basel/Knora/issues/978, the proplem arises, because the superclass introduced a very restrictive cardinality (1) and the subclass tried to loosen it. The ontology responder checks if cardinalities on the same property are loosened in the subclass. It should also check if the cardinalites inherited from superproperties are loosened, and forbid this. But if the cardinalities are the same, or even more restrictive, I don't see a problem.

I suggest to add something like the following check instead:

```
maybePropertyAndSubproperty match {
            case Some((basePropertyIri, propertyIri)) =>
                        val thisClassKnoraCardinality: KnoraCardinalityInfo = Cardinality.owlCardinality2KnoraCardinality(
                            propertyIri = propertyIri.toString,
                            owlCardinality = thisClassKnoraCardinalities(propertyIri)
                        )

                        val inheritableKnoraCardinality: KnoraCardinalityInfo = Cardinality.owlCardinality2KnoraCardinality(
                            propertyIri = basePropertyIri.toString,
                            owlCardinality = inheritableKnoraCardinalities(basePropertyIri)
                        )

                        if (!Cardinality.isCompatible(directCardinality = thisClassKnoraCardinality.cardinality, inheritableCardinality = inheritableKnoraCardinality.cardinality)) {
                            // No. Throw an exception.
                            throw BadRequestException(s"Class <${classDefWithAddedLinkValueProps.classIri.toOntologySchema(ApiV2WithValueObjects)}> has a cardinality on property <${basePropertyIri.toOntologySchema(ApiV2WithValueObjects)}> and on its subproperty <${propertyIri.toOntologySchema(ApiV2WithValueObjects)}>")
                        }
                        else
                          ()
                              
            case None => ()
        }
```

**Comment by benjamingeer** [comment]
> If you define an abstract concept that has an abstract property (e.g. a creation that can have 1 or more creators) and you are subclassing this, eg. having a sculpture that has 1 or more sculptors and a text that has 1 or more authors, then you would probably want to use this method.

I think you're not talking about the same thing. If you have a property `hasCreator` and subproperty `hasAuthor`, nothing prevents you from having a class `Text` with `minCardinality 1` on `hasAuthor`. Then your text can have multiple authors. This is how subproperties are supposed to work.

What is forbidden (and this was the key problem in #978) is to have a class `Text` with a cardinality on `hasAuthor` **and** a separate cardinality on `hasCreator`. This is not possible, and in my view it's illogical.

It's not possible because there is no way for the triplestore to check the two cardinalities separately.

Suppose you have a class with `cardinality 1` on `hasAuthor` **and** a separate `cardinality 1` on `hasCreator`. If your resource has a value for `hasAuthor`, but no value for `hasCreator`, then from the triplestore's point of view, you have satisfied **both** cardinalities, and the triplestore will not allow you to add a value for `hasCreator`. Your cardinality on `hasCreator` is therefore unusable.

And this is why the whole idea is illogical. It does not make sense to have a cardinality on one property that will actually be satisfied by another property.

To put it another way, it doesn't make sense to say that a text can have an author **and** a creator, and that those can be two different people. In the world of RDF, if you have an author, you already have a creator.

What would make sense would be if you had another subproperty of `hasCreator`, e.g. `hasEditor`. Then your `Text` class could have, say, `cardinality 1` on `hasCreator` and `cardinality 1` on `hasEditor`, and you would have no problem. The author and the editor would be two different people.

**Comment by mattssp** [comment]
OK. So your check allows inherited constraints on superproperties? That would resolve the creator case.

Still, there are lots of cases where there is no problem with having cardinalities on properties and subproperties. And there are lots of them in the nie-ine-ontologies. Most of the time, they have a minCardinality of 0, and there will never be a problem with this.

One example I am stumbling over is the following:

```
human:Creation
	a owl:Class;
	rdfs:comment """Something created by one or more persons."""@en;
	rdfs:subClassOf event:Output, cidoc:E71_Man-Made_Thing, [
		a owl:Restriction; owl:onProperty text:hasName; owl:minCardinality "0"^^xsd:nonNegativeInteger], [
		a owl:Restriction; owl:onProperty text:hasIdentifier; owl:minCardinality "0"^^xsd:nonNegativeInteger], [
		a owl:Restriction; owl:onProperty text:hasAlias; owl:minCardinality "0"^^xsd:nonNegativeInteger],  

[...]
```
An alias is also a name (which makes sense), there could be several different types of names, some of which are specified as Alias or whatnot. I don't see something illogical here. I agree that a resource should not have an Author and a seperate abstract Creator that is not the same person. But this is a special case.

There are more examples eg. texts having textreferences and footnotesreferences (beeing textreferences).

So I still think the check is to restrictive. It should at least allow the case where minCardinality is 0 (and thus no real constraint is given, just a possibility)

**Comment by benjamingeer** [comment]
> So your check allows inherited constraints on superproperties?

Yes. If `Creation` has `minCardinality 1` on `hasCreator`, and `Text` has `minCardinality 1` on `hasAuthor`, there is no problem. The constraint in the subclass can also be more restrictive (e.g. `cardinality 1`), but it cannot be less restrictive (because the constraint on the base property is inherited and must therefore be satisfied).

> An alias is also a name (which makes sense), there could be several different types of names, some of which are specified as Alias or whatnot. I don't see something illogical here.

If you have an alias, then you have a name. So it is redundant to say "I have an alias and a name."

It's as if I said, "I have a brother and a sibling." At best this is ambiguous. Am I talking about one person or two?

To put it another way: why does `Creation` have a cardinality on `hasAlias` at all? Isn't it sufficient to have the cardinality on `hasName`?

**Comment by mattssp** [comment]
What if it could have more than one name, but, say, just one shortcode (which is a name)? The only case where I see technical problems with this kind of thing is when the superproperty has cardinality=1 or a maxCardinality set. And I don't see why the system should restrict the modelling if there is no technical reason for it.

**Comment by benjamingeer** [comment]
> What if it could have more than one name, but, say, just one shortcode (which is a name)?

If you want to create values for `hasName` (not for the shortcode), Knora will require you to use `hasName` directly, not a subproperty like `hasFamilyName`. Is that really what you want? If you know that there are different types of names, you should have different properties for them, and use those, rather than use a vague base property. This is just good data modelling practice: be as specific as possible.

> And I don't see why the system should restrict the modelling if there is no technical reason for it.

I think enforcing clear modelling practices is a good thing. It makes data more maintainable and reusable.

**Comment by mattssp** [comment]
We are talking about a very abstract class in a shared ontology here. So you would enforce all subclasses to define all types of names they use, and forbid to  have just general names (creations have just names, not family names) and then special names like the short code. Or define something like OrdinaryName. Or an AbstractName superclass of Name and Shortcode (and if I search for Names I don't find the shortcode, because i would have to search for AbstractNames instead)? This may be good practice (I'm not sure, tho), but makes the ontology much less understandable.

**Comment by benjamingeer** [comment]
> We are talking about a very abstract class in a shared ontology here.

OK, so now we actually have just one very specific use case: an abstract base class with `minCardinality: 0` on a superproperty and a more restrictive cardinality on a subproperty.

> So you would enforce all subclasses to define all types of names they use, and forbid to have just general names (creations have just names, not family names) and then special names like the short code.

Here's one option:

- `hasShortcode rdfs:subPropertyOf hasName`
- `hasTitle rdfs:subPropertyOf hasName`
- `Creation`
  - `hasName minCardinality 0`
- `Text`
  - `hasTitle cardinality 1`
  - `hasShortcode cardinality 1`

Then you can search for `hasName`, and get results for all the subproperties.

If you want to specify that a `Creation` can have only one shortcode, here's another option:

- `hasShortcode rdfs:subPropertyOf hasName`
- `hasTitle rdfs:subPropertyOf hasName`
- `Creation`
  - `hasTitle minCardinality 0`
  - `hasShortcode cardinality 1`
- `Text`
  - `hasTitle cardinality 1`
  - `hasShortcode cardinality 1`

Both of these options seem clearer and more understandable to me than having `hasName` and `hasShortcode` in the same class, which would be ambiguous and apparently redundant.

I would also ask whether a shortcode is actually a name. Isn't it rather an ID, something machine-readable but not human-readable? If so, perhaps it should not be a subproperty of `hasName`.

**Comment by mattssp** [comment]
Look, the shortcode (which is for humans and is a name) is just an example. I am trying to use the ontologies developed by nie-ine (because I want to be compatible with them) and there are a lot of problems of this kind there, all due to this constraint. Apparently I will have to wait for Hans to resolve this compatability issue on his side (for now i adapted the constraint in my copy of knora, but this is not a long term solution).

**Comment by benjamingeer** [comment]
> Look, the shortcode (which is for humans and is a name) is just an example.

If you have a real use case that can't be better solved by improving the ontology design as I suggested, please let us know.

> I am trying to use the ontologies developed by nie-ine (because I want to be compatible with them) and there are a lot of problems of this kind there, all due to this constraint.

I think it would be helpful for NIE-INE to read the discussion that you and I have had here, and think about changing their ontologies in the ways I have suggested. If you or they can provide a **real** example where changing the class isn't a good solution, keeping in mind the points I've made, then we can talk about it again.
